### ISSUE TITLE
WebSocket Connection Getting Closed Abruptly

### ISSUE THREAD

**Issue by rogtrader** [polygon-io/issues#125]
**Client**
Which WebScoket client are using?

from polygon import RESTClient, WebSocketClient, STOCKS_CLUSTER

Polygon API Pythong Client (WebSocketClient)

**Issue**
A clear and concise description of the issue.

>>Connection closes abruptly after an hour or so.

**Expected Result**
>>Connection needs to be active for the trading session from 4am to 8pm est

**Screenshots**
Not Applicable

**Desktop (please complete the following information):**
 - OS: [e.g. macOS]
Python 3.6 - Linux CentOS 7

**Additional context**

**Comment by teamun1corn** [comment]
Its a bug on the client side that I havent figured out. Same bug most likely that is taking 2000+ messages down to 300 over time, then shortly there after, kills my connect.

**Comment by jrbell19** [comment]
When the connection gets closed, are you getting an error message/code?

**Comment by teamun1corn** [comment]
No must it looks like your forcing me to use wss protocol which would explain the sluggishness. 
![image](https://user-images.githubusercontent.com/13400134/109207101-5a4c7480-776e-11eb-9186-b69747764452.png)

**Comment by nickmachairas** [comment]
Hello, new to your platform, I am on the Stocks Developer plan. The webocket connection (A. * ) is highly unstable between 9:30am-10:30am EST and 3:30pm-4:00pm. Monday (3/8) was really bad, Tuesday (3/9) noticeably better but still dropping connections, today Wednesday (3/10) was absolutely horrible, see screenshot. I am on a stable fiber internet connection, very close to your datacenter and my system not nearly overloaded, even at spikes of ~3k mgs/sec for A. *  (rare). Outside of these windows the connection is stable. The pattern suggests that the issue is on your end, but I am not an expert. I get error 1006 when the connection drops.

Any ideas? Happy to work with you and troubleshoot. Thanks!

<img width="967" alt="polygon_disconnect" src="https://user-images.githubusercontent.com/15345135/110722524-6c073080-81e0-11eb-9767-5f7c158596b1.png">

**Comment by teamun1corn** [comment]
Do you have a Message Queue setup?

**Comment by nickmachairas** [comment]
Not yet. Speaking of MQ, any advice on producing websocket messages into Kafka? I am surprised by the lack of documentation on websockets, probably searching for the wrong thing.

Right now I am testing the polygon stream with a simple async python script. Average ~500 msg/sec is too low to cause backlog and disconnect on my end. This is the error message:

```
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
```

**Comment by teamun1corn** [comment]
I have a queue up and running. While my code is not optimized yet, my queue the other day, ran into 3 million messages waiting to be processed. So I would suggest getting a queue. From some of the forums I'm apart of, they suggest zeroMQ and from what they say it is faster than RabbitMQ.  Sounds like your code is pretty well optimized to handle 500 msg/sec. When I first started about three weeks ago, I could barely handle 100 message per second. Now I am filtering and processing each message on average without writing to an array in about 60 -110 microseconds, and I am fairly certain I can optimize that down to about 30-80 mics in javascript!  Your issue is because you dont have a queue, and possibly running via a single thread, you just cant keep up. There is something called a latency median of 99% - 50%. You happen to be running into the 99% - 50% and then your filling up the server side buffer because you filled up your TCP window and as a client you cant "ack" the messages fast enough to the polygon servers, so in a  sense, you are throttling yourself because you have not quite optimized your operation. Look into multi threading your operation, getting a queue to open up the TCP window and you should have all the data you need. 

And I am very thankful the polygon staff was patient with all my questions. 

I didnt read this whole thing but this is where your current problem lies.

[ https://engineering.linkedin.com/performance/who-moved-my-99th-percentile-latency#:~:text=A%2099th%20percentile%20latency%20of,experience%2030%20ms%20of%20delay.&text=This%20means%20that%20the%201,of%201%20million%20client%20requests.](url)

And this is how you solve it:

Bottom line -- 

1. - Get a queue up and running
2. - multi-core/multi threaded script,
3. -   Optimize your functions,
4. -   keep your TCP windows wide open by doing 1-3

**Comment by teamun1corn** [comment]
![RAbbit MQ](https://user-images.githubusercontent.com/13400134/110802750-17959c80-8244-11eb-93e7-81fa162e4f94.PNG)

**Comment by nickmachairas** [comment]
Impressive! Thanks for taking the time to post. Agreed, MQ is the way to go and it's were I'm heading. Hopefully a (somewhat) performant managed service that I can set up and forget. But I've got a lot to learn so I am taking it one step at a time. When I mentioned a simple async script, I mean very simple. No processing, I am just storing all A. * messages to a TimescaleDB instance that I've benchmarked to handle 100,000 msg/sec without a sweat. So the 500 msg/sec drop outs are troubling. I want to find the root of this problem before setting up larger infrastructure.

The lack of documentation on programmatically working with the websocket streams is a bit frustrating but I like polygon, so far. I am glad they helped you, I hope they can help me too. I am trying to get more ideas on async python scripts by reading into the Alpaca docs and code.

You went with RabbitMQ? How did you set up the incoming stream? Are you getting any disconnects? Thanks!

**Comment by teamun1corn** [comment]
Well the root cause is your TCP window is not staying open. If you do a wire capture on your wss connection to Polygon.io, you'll consistently see that the TCP window was closed. Meaning, your client side is not responding to the "ACK" messages that are being sent from the server. In a normal conversation with a human, we start out with, "hello" and the other person says in return, "Hi, how are you?". If the person doesn't respond, why waste your time trying to maintain a conversation with them if they are not processing or responding to your friendly verbal request in a timely fashion? Same with a TCP connection. Computers like to be heard and they like to speak, whether a server or a client and I guess they are kind of needy too.. 

Server: "Did you accept my message?"
Client: "Yes, I accpeted your message, please send more"
Server: "Ok, I'm going send you A LOT MORE, but only if you accept them!" 

12K msg per second later without a queue
Client: "crickets..................."
Server: (Server comes to a logical conclusion) "Hmm, they must no longer accept or handle all my messages, whatever time to move on, there is more fish in the sea for me to send messages to"

It took me several energy drinks, hemming and hawing and finally I followed a tutorial on rabbitMQ and bam, three minutes later after staring at my screen for 12 hours looking for the motivation to follow the tutorial, I had a queue full of 3 million message just patiently waiting to be processed. The issue was not my provider, it was me. I was throttling myself!  I am also learning. I just started getting serious about programming back in August. Now I'm processing messages in microseconds on a laptop with a residential connection. When my script is out of dev stage, moving to a vps with 16 cores. 

Am I getting any disconnects? No. I'm averaging 1200 -2000 messages currently on my laptop, using Javascript, with only 0-60 messages enqueue. I am also just asking for what I can handle. 


Oh I searched hours within my Websocket library trying to find a work around, as I like to call it, "My client side throttle". It really is just as simple as following a tutorial on an opensource enterprise Messaging Queue. Then getting really creative on how your handling the data, and use of multi-cores. 

I still need to figure out how to put all the pieces together but I am successfully processing live data against historical data in micro seconds! 

How I setup the incoming stream is, I took the WSS feed from Polygon, and fed the data via an event Emitter into my queue script. This is what you  call your publisher script. Seriously is only about 80 lines of code. (follow a tutorial)

Then you have a consumer script. This is where you pull from the queue, that code is roughly 20 lines. (follow a tutorial)


Hopefully I answered some of your questions and you can see some light at the end of the tunnel. Good luck! Let me know if you have any more questions! I guess during Market hours my latency from processing time increase.. It will be faster! 

![image](https://user-images.githubusercontent.com/13400134/110813170-ddc99380-824d-11eb-93e5-58fcd2a9adb4.png)

**Comment by teamun1corn** [comment]
I got sorting the list down to .005 mics with 2500 items in the list. Using NodeJS V8!  Will know more at market open if live data can still reference the memory location fairly certain.  If I can access it now with two different functions, I can take the wss feed and access it.   
![image](https://user-images.githubusercontent.com/13400134/110917758-e36fb980-82df-11eb-9051-13c7fd56a8c0.png)

**Comment by nickmachairas** [comment]
Thanks again for all the feedback, I am glad it is going great for you!

I am still plagued by disconnects. I migrated all my code to Google Cloud, set up a publishing routine to pubsub but it still keeps dropping the connection during peak hours. Which means that my local network was not the issue.

I think it is Python. Here's what I do:

```python
async def stream():
    async with websockets.connect(POLYGON_URI, ssl=True) as websocket:
        await websocket.send(
            '{"action":"auth","params":"' + POLYGON_KEY + '"}')
        await websocket.send(
            '{"action":"subscribe","params":"' + POLYGON_STREAM + '"}')

        greet = await websocket.recv()
        auth = await websocket.recv()
        sub = await websocket.recv()
        print(greet, '\n', auth, '\n', sub, '\n')

        async for messages in websocket:
            # ... publish to XYZ
```

The reason why I went with websockets instead of polygon's client (basically a port of websocket_client's app example) is I did some benchmarking and with asyncio the process runs on more threads.

Super simple IMO but something's going causing the connection to drop. I would ***really*** appreciate some feedback from Polygon. In the meantime I guess I'll try nodejs.

**Comment by teamun1corn** [comment]
I went the same route you did. Got a google VPS ran my code and got disconnected_**### I already did all the homework for you.**_ Do a wire capture on your wss connection. You will see your tcp window is full, meaning you cant accept any more packets until your client side websocket buffer allows for more room. Get a queue! Today on my network I ran all day pulling down quote and trade data, messages in the 12k range all day! My only issue now is how to pull it out of the queue fast enough. I can processes it quite quickly but I'm doing something wrong with how im accessing the queue and or asking the queue for the data. 

Also when I choose turn on my VPS, which has 2500 Mb/s down and a few hops from their datacenter, without a queue, it could only stay in the fight for only a minute at most during highly congested times. Why??? The amount of data they are sending per second is insane! And your Websocket library's buffer sure as hell cant handles the on slaught of data your happily paying for.  Have you done the math from bits to bytes per message per second? That right there will tell you how big your client side WSS buffer will need to be in order to handle the data.  How big is your client side Websocket Buffer? 1 gig?  in 3 seconds you filled that buffer up, in 4 your got disconnect. 

I am suspecting setting ping_timeout = None did nothing. 

Can you tell me from your script, how many message you received and processed against what their dashboard is saying? 
Can you tell me how many message you were receiving when you lost connection from your scripts perspective and then from polygons dashboard?

**Comment by teamun1corn** [comment]
If you have any questions on nodejs, I would be happy to answer them. I am no expert though, I just like solving problems I cant solve today with brute force tomorrow.

**Comment by teamun1corn** [comment]
FYI, I am fairly certain also that there sending for T.* a rate that is instantaneous rather than per second. Noticing my exchange is publishing 13k +- 2k - 10k per second depending on market congestion. So messages on the dashboard != per second.

**Comment by nickmachairas** [comment]
That is true. But I am not even close to that volume, there are bigger problems to solve, namely keeping the connection alive. Python failed, Golang failed and all I get from Polygon is silence. Absolutely horrible customer service.

**Comment by teamun1corn** [comment]
https://www.youtube.com/watch?v=TkCSr30UojM 

Do a wire capture on your wss feed. Try to limit the amount of tcp connections you have closing all extra processes and services that are leveraging an internet connection. Then start your script. At some point when your connection starts to drop or tapper off, you should see colors on the wireshark capture that are red and black.. Find that error and google it.  That will give you the clues you need to isolate where the problem lies. If you never have a tcp "full window" error pop up, then it is a server side issue. If you have a tcp full window, then its a client side issue. 

If you have any more questions let me know. I hope I was helpful. To solve this problem, I connected all the dots on my own and spent hours digging through the ws library on the client side.

**Comment by nickmachairas** [comment]
Thank you so much, I will definitely try this and report back. Next week though, when trading activity resumes. Thanks again!

**Comment by teamun1corn** [comment]
https://stackoverflow.com/questions/32585481/wireshark-rst-against-tcp-zero-window 

I'm fairly certain your issue will be on the client side. Today my connection dropped, but... but I ended up exhausting all my memory via my publisher. Any recommendations on fast db?

**Comment by teamun1corn** [comment]
Btw check out these results.. Clustered queues, clustered consumers... Now I need a database. 

![image](https://user-images.githubusercontent.com/13400134/111856305-c5b1de00-88f7-11eb-8761-6b3543dacfd3.png)

**Comment by nickmachairas** [comment]
And you're reading all those messages in with Javascript?! Impressive! How I cannot get a stable connection with 600 msg/sec and Go coroutines is a mystery. I'll keep digging...

My recommendation is TimescaleDB, it's a postgres extension. You'll get very good performance on a single node, even better with multiple nodes. And also get to enjoy all the postgres goodies.

**Comment by teamun1corn** [comment]
Yes with JavaScript. I'm still learning. It is probably the worst language to build this thing in. But I'm not trying to build a HFT. I'm trying to build a tool to inform. After I have a few wins from the informing algo, then it will be time to port it over to a new language that is more adapt to speed.   I'll work on getting Timescale implemented today. 

Your goal here is to pull the messages off the wss connection as fast as possible,, freeing up memory and your client side buffer, that way we can identify if there are other bottlenecks prior to and well before you app. Also, I'm not using any async functions in my script. The thought is with live data I should be asyncing awaiting for any of it. Add the the database with bring in some complexity to latency though. May need an app layer queue inside my script now.

BTW rabbitMQ has a slack channel with very fast response time when you have questions. So I would suggest RabbitMQ as a messaging provider also if this is your first implementation of a queue.

**Comment by teamun1corn** [comment]
https://www.youtube.com/watch?v=0ZsGZwhmkB4&t=2s

**Comment by nickmachairas** [comment]
@teamun1corn I got an email that you posted here but I can't see your post. Nice work on your system! I recommended TimescaleDB above, and now I will recommend Golang, it has been such an amazing tool for me. The problem I had before was that I was waiting for confirmation from the MQ that the message was successfully published, just a single line of code. This is what happened once I removed it:

![Screen Shot 2021-04-13 at 5 06 41 PM](https://user-images.githubusercontent.com/15345135/114622363-d990fb80-9c7b-11eb-8ae0-39126920c138.png)

On GCP, Pub/Sub, BigQuery for long term storage and a brave n2d-highcpu-4 instance (4 vCPUs AMD Epyc Rome, 4 GB memory). A lot of arguments can be made about latency and how the whole operation could run better/faster on a different system. But I am okay for now, learned a ton, still learning and managed services mean fewer failure points and headaches. It is very much a work in progress with code that is far from optimal.

If only Polygon had better docs for the API clients and was more responsive... They have a unique product though so I can't be too mad ;)

**Comment by teamun1corn** [comment]
Aahah nice! I think your numbers are crushing mine now! I now have some more work to do!

**Comment by teamun1corn** [comment]
Awesome work man!

**Comment by nickmachairas** [comment]
Thanks! I am not running any computations yet, this is a next step and will happen on a separate VM with subscribers only.

The graphs are part of the GCP monitoring suite, I didn't do much to produce them. I also didn't do much to split the work among the cores, just set up separate coroutines on Go. I am surprised that the n2d VM with "a base frequency of 2.25 GHz, an effective frequency of 2.7 GHz, and a max boost frequency of 3.3 GHz" and just 2 cores (4 vCPUs) can handle up to 25-30,000 msgs/sec. Again, not optimal code at all, I could do much better. Your 3950x should do so *much* better with 16 cores (roughly equivalent to 32 vCPU) and perhaps more importantly stock 3.5GHz base and 4.7GHz boost frequency, higher if overclocked. I am considering building a rig but for now I opted for splitting the work among multiple small VMs on GCP and see how it goes.

They say Javascript is the wild kid after whom you might need to clean up sometimes. Try Go!

**Comment by teamun1corn** [comment]
I think you're right as everyone has said, I need to change languages! I climbed this mountain disabled with javascript and made it to the summit! Am I ready to re-write this whole thing?!?!? Probably. I'm new to a lot of things, including programming in general and processing tons of data as you have said, there are many ways I can optimize.  Leaving the red headed step child, Javascript, is one of them. 

Anyways glad to see you made it!


here's my discord if you ever want to connect. I'm normally up at all hours. 

https://discord.gg/ueXj86yj
